packages/astro/src/cli/check/index.ts

```diff
@@ -1,101 +1,360 @@
+ */
+export async function check(
+	settings: AstroSettings,
+	{ logging, flags }: CheckPayload
+): Promise<AstroChecker> {
+	const checkFlags = parseFlags(flags);
+	if (checkFlags.watch) {
+		info(logging, 'check', 'Checking files in watch mode');
+	} else {
+		info(logging, 'check', 'Checking files');
+	}
 
+	const { syncCli } = await import('../../core/sync/index.js');
+	const root = settings.config.root;
 	const require = createRequire(import.meta.url);
-	let checker = new AstroCheck(
+	const diagnosticChecker = new AstroCheck(
 		root.toString(),
-		require.resolve('typescript/lib/tsserverlibrary.js', { paths: [root.toString()] })
+		require.resolve('typescript/lib/tsserverlibrary.js', {
+			paths: [root.toString()],
+		})
 	);
-	const filesCount = await openAllDocuments(root, [], checker);
 
-	let diagnostics = await checker.getDiagnostics();
+	return new AstroChecker({
+		syncCli,
+		settings,
+		fileSystem: fs,
+		logging,
+		diagnosticChecker,
+		isWatchMode: checkFlags.watch,
+	});
+}
 
-	spinner.succeed();
+type CheckerConstructor = {
+	diagnosticChecker: AstroCheck;
 
-	let result: Result = {
-		errors: 0,
-		warnings: 0,
-		hints: 0,
-	};
+	isWatchMode: boolean;
 
-	diagnostics.forEach((diag) => {
-		diag.diagnostics.forEach((d) => {
-			console.log(printDiagnostic(diag.fileUri, diag.text, d));
+	syncCli: (settings: AstroSettings, options: SyncOptions) => Promise<ProcessExit>;
 
-			switch (d.severity) {
-				case DiagnosticSeverity.Error: {
-					result.errors++;
-					break;
-				}
-				case DiagnosticSeverity.Warning: {
-					result.warnings++;
-					break;
-				}
-				case DiagnosticSeverity.Hint: {
-					result.hints++;
-					break;
-				}
+	settings: Readonly<AstroSettings>;
+
+	logging: Readonly<LogOptions>;
+
+	fileSystem: typeof fsMod;
+};
+
+/**
+ * Responsible to check files - classic or watch mode - and report diagnostics.
+ *
+ * When in watch mode, the class does a whole check pass, and then starts watching files.
+ * When a change occurs to an `.astro` file, the checker builds content collections again and lint all the `.astro` files.
+ */
+export class AstroChecker {
+	readonly #diagnosticsChecker: AstroCheck;
+	readonly #shouldWatch: boolean;
+	readonly #syncCli: (settings: AstroSettings, opts: SyncOptions) => Promise<ProcessExit>;
+
+	readonly #settings: AstroSettings;
+
+	readonly #logging: LogOptions;
+	readonly #fs: typeof fsMod;
+	#watcher?: FSWatcher;
+
+	#filesCount: number;
+	#updateDiagnostics: NodeJS.Timeout | undefined;
+	constructor({
+		diagnosticChecker,
+		isWatchMode,
+		syncCli,
+		settings,
+		fileSystem,
+		logging,
+	}: CheckerConstructor) {
+		this.#diagnosticsChecker = diagnosticChecker;
+		this.#shouldWatch = isWatchMode;
+		this.#syncCli = syncCli;
+		this.#logging = logging;
+		this.#settings = settings;
+		this.#fs = fileSystem;
+		this.#filesCount = 0;
+	}
+
+	/**
+	 * Check all `.astro` files once and then finishes the operation.
+	 * @returns {Promise<CheckResult>}
```

Comment:
I think we can remove this `@return` type as it's already annotated in TypeScript below.